### PR TITLE
Process `snapshot` tokens

### PR DESCRIPTION
**Pull request by davidomid**
Closes: https://github.com/github/hosted-runners/issues/526 

Summary of changes: 

- Added deserialisation of a new `TemplateToken` called `snapshot` to `AgentJobRequestMessage`. 
- Added evaluation of `snapshot` from `TemplateToken` to a new `Snapshot` type. 
- Added a new `SnapshotOperationProvider` service, responsible for processing `Snapshot` objects. 
- Updated `JobExtension` to evaluate a `Snapshot` if one is present in the job template, then register a new post-job step for invoking `SnapshotOperationProvider` with it. 
- Added new L0 tests for `JobExtension` and `SnapshotOperationProvider`. 
- Updated existing tests for regression.

### CODE REVIEW

File: src/Runner.Worker/JobExtension.cs
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs:408]: Can you confirm the order? I would imagine the post steps run in the following order:
- Snapshot
- Completed hook
- Job cleanup

File: src/Runner.Worker/JobExtension.cs
Comment (ericsciple) [inline src/Runner.Worker/JobExtension.cs:20]: curious is this used?

**Review by ericsciple — DISMISSED** [review]
Thanks! This makes sense to me. Although I would like to get a second review from someone who works in the runner more recently.

File: src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateConverter.cs
Comment (TingluoHuang) [inline src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateConverter.cs:349]: are these changes copied from the service?
we we expect the golang parser handles this?

File: src/Sdk/DTPipelines/workflow-v1.0.json
Comment (TingluoHuang) [inline src/Sdk/DTPipelines/workflow-v1.0.json:75]: is this schema change coming from the actions-service or the parser?
i am trying to find the source and make sure they match each other.

File: src/Runner.Worker/JobExtension.cs
Comment (TingluoHuang) [inline src/Runner.Worker/JobExtension.cs:409]: i think this needs to move up before we calling `// Register Job Completed hook if the variable is set`

basically, snapshot first, then `job_complete_hook`

File: src/Runner.Worker/SnapshotOperationProvider.cs
Comment (TingluoHuang) [inline src/Runner.Worker/SnapshotOperationProvider.cs:27]: you can use `StringUtil.ConvertToJson()` instead of `JsonConvert.SerializeObject`

File: src/Runner.Worker/SnapshotOperationProvider.cs
Comment (TingluoHuang) [inline src/Runner.Worker/SnapshotOperationProvider.cs:20]: is the runner root directory the right place for putting this file?

or the WellKnownDirectory.Temp might be better since it get cleanup before each job.

File: src/Runner.Worker/SnapshotOperationProvider.cs
Comment (TingluoHuang) [reply src/Runner.Worker/SnapshotOperationProvider.cs:27]: you can also just use `IOUtil.SaveObject()` to do the josn serialization and write to disk in one line.

File: src/Runner.Worker/JobExtension.cs
Comment (davidomid) [reply src/Runner.Worker/JobExtension.cs:20]: Apologies, it was used originally before I migrated the logic to `SnapshotOperationProvider` and I forgot to remove the reference.

File: src/Sdk/DTPipelines/workflow-v1.0.json
Comment (davidomid) [reply src/Sdk/DTPipelines/workflow-v1.0.json:75]: It was added to the C# parser in this PR: https://github.com/github/actions-workflow-parser/pull/406

The Go parser hasn't been updated yet, that's happening in a separate batch: https://github.com/github/hosted-runners/issues/528

File: src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateConverter.cs
Comment (davidomid) [reply src/Sdk/DTPipelines/Pipelines/ObjectTemplating/PipelineTemplateConverter.cs:349]: I **_think_** the answer to these questions is similar to this comment: https://github.com/actions/runner/pull/3135/files#r1489319642

But please let me know if you need additional information. 

For context, we're unsure at this stage if we'll need expression support at some point and @ericsciple advised that it made sense to send the entire `TemplateToken` down to the runner and process it there instead of sending an evaluated `Snapshot` object. This was recommended as we would have the ability to check for expression tokens in the future. I'm happy to make changes if this wasn't the optimal approach though.

File: src/Runner.Worker/SnapshotOperationProvider.cs
Comment (davidomid) [reply src/Runner.Worker/SnapshotOperationProvider.cs:27]: Thanks for the advice! 

I opted for just using `IOUtil.SaveObject`. Originally I thought it made sense to have the operations separated as I was logging the snapshot object JSON afterwards but I ended up removing that as it seemed redundant. The behaviour should be adequately tested by the L0 tests.

File: src/Runner.Worker/JobExtension.cs
Comment (davidomid) [reply src/Runner.Worker/JobExtension.cs:408]: Thanks, I think you're right. I'll go ahead and move that up.

File: src/Runner.Worker/JobExtension.cs
Comment (davidomid) [reply src/Runner.Worker/JobExtension.cs:407]: @TingluoHuang pointed out that the order doesn't seem correct, so I've addressed that concern here: https://github.com/actions/runner/pull/3135/files#r1486509000

**Review by ericsciple — APPROVED** [review]
Thanks! This makes sense to me. Will defer to Ting whether he has any remaining concerns.

**Pull request merged**
